timestep.c

```diff
@@ -772,41 +751,106 @@ void rebuild_activelist(void)
  * function will drift to this moment, generate an output, and then
  * resume the drift.
  */
-inttime_t find_next_kick(inttime_t Ti_Current)
+inttime_t find_next_kick(inttime_t Ti_Current, int minTimeBin)
 {
-    /* Note that on startup, P[i].TimeBin == 0 for all particles,
-     * all bins except the zeroth are inactive and so we return 0 from this function.
-     * This ensures we run the force calculation for the first timestep.*/
-    /* find the smallest active bin*/
-    int n;
-    for(n = 0; n < TIMEBINS; n++)
-    {
-        if(TimeBinCount[n] > 0)
-            break;
-    }
-    MPI_Allreduce(MPI_IN_PLACE, &n, 1, MPI_INT, MPI_MIN, MPI_COMM_WORLD);
     /* Current value plus the increment for the smallest active bin. */
-    return Ti_Current + dti_from_timebin(n);
+    return Ti_Current + dti_from_timebin(minTimeBin);
 }
 
 /* mark the bins that will be active before the next kick*/
 int update_active_timebins(inttime_t next_kick)
 {
     int n;
-    int NumForceUpdate = TimeBinCount[0];
-
     for(n = 1, TimeBinActive[0] = 1; n < TIMEBINS; n++)
     {
         int dti_bin = (1 << n);
-
 /*         message(0, "kick: %d Tbin %d dti_bin %d ACTIVE %d\n", next_kick, n, dti_bin, next_kick % dti_bin == 0); */
         if((next_kick % dti_bin) == 0)
-        {
             TimeBinActive[n] = 1;
-            NumForceUpdate += TimeBinCount[n];
-        }
         else
             TimeBinActive[n] = 0;
     }
-    return NumForceUpdate;
+    return 0;
+}
+
+/*! This routine writes one line for every timestep.
+ * FdCPU the cumulative cpu-time consumption in various parts of the
+ * code is stored.
+ */
+void print_timebin_statistics(int NumCurrentTiStep)
+{
+    double z;
+    int i;
+    int64_t tot = 0, tot_type[6] = {0};
+    int64_t tot_count[TIMEBINS] = {0};
+    int64_t tot_count_type[6][TIMEBINS] = {0};
+    int64_t tot_num_force = 0;
+
+    for(i = 0; i < 6; i ++) {
+        sumup_large_ints(TIMEBINS, TimeBinCountType[i], tot_count_type[i]);
+    }
+
+    for(i = 0; i<TIMEBINS; i++) {
+        int j;
+        for(j=0; j<6; j++)
+            tot_count[i] += tot_count_type[j][i];
+        if(is_timebin_active(i))
+            tot_num_force += tot_count[i];
+    }
+
+    /* let's update Tot counts in one place tot variables;
+     * at this point there can still be holes in SphP
+     * because rearrange_particle_squence is not called yet.
+     * but anywaysTotN_sph variables are not well defined and
+     * not used any places but printing.
+     *
+     * we shall just say they we sync these variables right after gravity
+     * calculation in every timestep.
+     * */
+
+    char extra[1024] = {0};
+
+    if(is_PM_timestep(All.Ti_Current))
+        strcat(extra, "PM-Step");
+
+    z = 1.0 / (All.Time) - 1;
+    message(0, "Begin Step %d, Time: %g, Redshift: %g, Nf = %014ld, Systemstep: %g, Dloga: %g, status: %s\n",
```

Comment:
I've been thinking it would be nice to see the integer step, especially now that it is consistent between runs.